### PR TITLE
ext_authz: add an option to allow buffering requests on per-route filer

### PR DESCRIPTION
**Pull request by agrawroh**
## Background

Today, [ExtAuthZ Per-Router filter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthzperroute) only allows disabling the behavior for buffering the request body when it's enabled on the main filter in the HCM filter-chain. We have an opposite use-cases where we don't want to buffer the body from the routes except a handful of routes. 

## Changes

This PR switches the `disable_request_body_buffering` option which is part of `check_settings` to a one of and additionally let users specify `with_request_body` [[Ref]](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#envoy-v3-api-field-extensions-filters-http-ext-authz-v3-extauthz-with-request-body). This serves as a per-route override to explicitly enable the request body buffering behavior.

Commit Message: add an option to allow buffering requests on per-route filer
Additional Description: add a new field to let `with_request_body` be passed from the per-route ExtAuthZ filter
Risk Level: Low
Testing: Added Tests
Docs Changes: Yes
Release Notes: Added
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto [hunk 1]
```diff
@@ -382,7 +382,16 @@ message CheckSettings {
...
+
+    // When set to true, disable the configured :ref:`with_request_body
+    // <envoy_v3_api_field_extensions.filters.http.ext_authz.v3.ExtAuthz.with_request_body>` for a specific route.
+    bool disable_request_body_buffering = 2;
+
+    // Enable or override request body buffering, which is configured with
+    // :ref:`with_request_body <envoy_v3_api_field_extensions.filters.http.ext_authz.v3.ExtAuthz.with_request_body>`
+    // for a specific route.
+    BufferSettings with_request_body = 3;
+  }
```
Comment (wbpcode) [inline api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:396]: Please don't move the `disable_request_body_buffering` to a oneof directly. You can add a new `with_request_body`  field only and document it clearly only one field could be set or valid clearly.

Comment (agrawroh) [reply api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto]: Removed the `oneof` and added a validation check instead.

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc [hunk 2]
```diff
@@ -2756,6 +2756,171 @@ TEST_P(HttpFilterTestParam, NoCluster) {
...
+  data_.add(buffer3.toString());
+
+  Buffer::OwnedImpl buffer4("more data after watermark is set is possible");
+  EXPECT_EQ(Http::FilterDataStatus::StopIterationAndWatermark, filter_->decodeData(buffer4, true));
+
+  EXPECT_EQ(Http::FilterTrailersStatus::StopIteration, filter_->decodeTrailers(request_trailers_));
+}
+
+// Checks that the per-route filter can override the check_settings set on the main filter.
+TEST_F(HttpFilterTest, PerRouteCheckSettingsOverrideWorks) {
```
Comment (ggreenway) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc:2861]: This test is identical to the previous test except for the name

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall; some test issues to cleanup

/wait

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc [hunk 3]
```diff
@@ -2756,6 +2756,171 @@ TEST_P(HttpFilterTestParam, NoCluster) {
...
+                            "Invalid configuration for check_settings. Only one of "
+                            "disable_request_body_buffering or with_request_body can be set.");
+}
+
+// Checks that the per-route filter can override the check_settings set on the main filter.
+TEST_F(HttpFilterTest, PerRouteCheckSettingsWorks) {
+  InSequence s;
+
+  envoy::extensions::filters::http::ext_authz::v3::ExtAuthzPerRoute settings;
+  FilterConfigPerRoute auth_per_route(settings);
```
[the remark this answers is not in the document]
Comment (agrawroh) [reply test/extensions/filters/http/ext_authz/ext_authz_test.cc:2801]: Done, Thanks!

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc (hunk 2 above)
Comment (agrawroh) [reply test/extensions/filters/http/ext_authz/ext_authz_test.cc:2861]: Sorry, the previous test was meant to be slightly different where we don't initially specify the `with_request_body` on the main filter. I fixed it.

**Review by ggreenway — APPROVED** [review]

**Pull request merged**